File: src/dionysus/audit.py
"""Service helpers for recording append-only audit events."""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import Any

from sqlalchemy.orm import Session

from dionysus.models.audit import AuditLogEvent

_REDACTED = "[REDACTED]"
_SENSITIVE_KEYS = {
    "authorization",
    "client_secret",
    "password",
    "secret",
    "token",
    "access_token",
    "refresh_token",
}
_LARGE_SENSITIVE_KEYS = {
    "raw_report",
    "report",
    "report_payload",
    "stack",
    "stack_trace",
    "stacktrace",
    "traceback",
}


def record_audit_event(
    session: Session,
    *,
    event_type: str,
    actor_principal_type: str | None = None,
    actor_principal_id: str | None = None,
    actor_display: str | None = None,
    target_type: str | None = None,
    target_id: str | None = None,
    project_id: str | None = None,
    ip_address: str | None = None,
    user_agent: str | None = None,
    metadata: Mapping[str, Any] | None = None,
) -> AuditLogEvent:
    """Record one sanitized audit event in the caller's transaction.

    Args:
        session: SQLAlchemy session used for persistence.
        event_type: Stable dotted event type describing what happened.
        actor_principal_type: Optional principal type for the actor.
        actor_principal_id: Optional principal ID for the actor.
        actor_display: Optional human-readable actor name.
        target_type: Optional domain type affected by the event.
        target_id: Optional domain ID affected by the event.
        project_id: Optional project ID associated with the event.
        ip_address: Optional request client IP address.
        user_agent: Optional request user agent.
        metadata: Optional event-specific safe metadata.

    Returns:
        The pending ``AuditLogEvent`` model.

    Raises:
        ValueError: If ``event_type`` is blank.
    """

    normalized_event_type = event_type.strip()
    if not normalized_event_type:
        raise ValueError("event_type is required")

    event = AuditLogEvent(
        event_type=normalized_event_type,
        actor_principal_type=str(actor_principal_type) if actor_principal_type else None,
        actor_principal_id=actor_principal_id,
        actor_display=actor_display,
        target_type=target_type,
        target_id=target_id,
        project_id=project_id,
        ip_address=ip_address,
        user_agent=user_agent,
        metadata_json=_sanitize_metadata(metadata or {}),
    )
    session.add(event)
    return event


def _sanitize_metadata(metadata: Mapping[str, Any]) -> dict[str, Any]:
    """Return metadata with obvious sensitive or oversized values redacted."""

    sanitized: dict[str, Any] = {}
    for key, value in metadata.items():
        key_text = str(key)
        if _is_sensitive_key(key_text):
            sanitized[key_text] = _REDACTED
            continue
        sanitized[key_text] = _sanitize_value(value)
    return sanitized


def _sanitize_value(value: Any) -> Any:
    if isinstance(value, Mapping):
        return _sanitize_metadata(value)
    if isinstance(value, str | int | float | bool) or value is None:
        return value
    if isinstance(value, Sequence) and not isinstance(value, str | bytes | bytearray):
        return [_sanitize_value(item) for item in value]
    return str(value)


def _is_sensitive_key(key: str) -> bool:
    normalized = key.casefold().replace("-", "_").replace(" ", "_")
    if normalized in _LARGE_SENSITIVE_KEYS:
        return True
    return any(part in _SENSITIVE_KEYS for part in normalized.split("_"))

File: backend/internal/audit/audit.go
package audit

import (
	"context"
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"reflect"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/invacuation/dionysus/backend/internal/db/dbgen"
)

const redacted = "[REDACTED]"

var sensitiveKeys = map[string]bool{
	"authorization": true,
	"client_secret": true,
	"password":      true,
	"secret":        true,
	"token":         true,
	"access_token":  true,
	"refresh_token": true,
}

var largeSensitiveKeys = map[string]bool{
	"raw_report":     true,
	"report":         true,
	"report_payload": true,
	"stack":          true,
	"stack_trace":    true,
	"stacktrace":     true,
	"traceback":      true,
}

type Event struct {
	Type               string
	ActorPrincipalType *string
	ActorPrincipalID   *string
	ActorDisplay       *string
	TargetType         *string
	TargetID           *string
	ProjectID          *string
	IPAddress          *string
	UserAgent          *string
	Metadata           map[string]any
	Now                time.Time
}

func RecordEvent(ctx context.Context, conn *sql.DB, event Event) (dbgen.AuditLogEvent, error) {
	eventType := strings.TrimSpace(event.Type)
	if eventType == "" {
		return dbgen.AuditLogEvent{}, errors.New("event_type is required")
	}
	now := event.Now
	if now.IsZero() {
		now = time.Now().UTC()
	}
	metadataJSON, err := json.Marshal(sanitizeMetadata(event.Metadata))
	if err != nil {
		return dbgen.AuditLogEvent{}, err
	}
	return dbgen.New(conn).CreateAuditLogEvent(ctx, dbgen.CreateAuditLogEventParams{
		ID:                 uuid.NewString(),
		EventType:          eventType,
		ActorPrincipalType: nullString(event.ActorPrincipalType),
		ActorPrincipalID:   nullString(event.ActorPrincipalID),
		ActorDisplay:       nullString(event.ActorDisplay),
		TargetType:         nullString(event.TargetType),
		TargetID:           nullString(event.TargetID),
		ProjectID:          nullString(event.ProjectID),
		IpAddress:          nullString(event.IPAddress),
		UserAgent:          nullString(event.UserAgent),
		MetadataJson:       string(metadataJSON),
		CreatedAt:          now.UTC(),
	})
}

func sanitizeMetadata(metadata map[string]any) map[string]any {
	sanitized := map[string]any{}
	for key, value := range metadata {
		keyText := key
		if isSensitiveKey(keyText) {
			sanitized[keyText] = redacted
			continue
		}
		sanitized[keyText] = sanitizeValue(value)
	}
	return sanitized
}

func sanitizeValue(value any) any {
	if value != nil {
		reflectedValue := reflect.ValueOf(value)
		if (reflectedValue.Kind() == reflect.Pointer || reflectedValue.Kind() == reflect.Interface) && reflectedValue.IsNil() {
			return nil
		}
	}
	switch typed := value.(type) {
	case map[string]any:
		return sanitizeMetadata(typed)
	case []any:
		items := make([]any, 0, len(typed))
		for _, item := range typed {
			items = append(items, sanitizeValue(item))
		}
		return items
	case string, int, int64, float64, bool, nil:
		return typed
	default:
		if reflected := sanitizeReflectedValue(reflect.ValueOf(value)); reflected != nil {
			return reflected
		}
		return fmt.Sprint(typed)
	}
}

func sanitizeReflectedValue(value reflect.Value) any {
	if !value.IsValid() {
		return nil
	}
	for value.Kind() == reflect.Pointer || value.Kind() == reflect.Interface {
		if value.IsNil() {
			return nil
		}
		value = value.Elem()
	}
	switch value.Kind() {
	case reflect.Map:
		if value.Type().Key().Kind() != reflect.String {
			return nil
		}
		sanitized := map[string]any{}
		for _, key := range value.MapKeys() {
			keyText := key.String()
			if isSensitiveKey(keyText) {
				sanitized[keyText] = redacted
				continue
			}
			sanitized[keyText] = sanitizeValue(value.MapIndex(key).Interface())
		}
		return sanitized
	case reflect.Slice, reflect.Array:
		items := make([]any, 0, value.Len())
		for i := 0; i < value.Len(); i++ {
			items = append(items, sanitizeValue(value.Index(i).Interface()))
		}
		return items
	default:
		return nil
	}
}

func isSensitiveKey(key string) bool {
	normalized := strings.NewReplacer("-", "_", " ", "_").Replace(strings.ToLower(key))
	if largeSensitiveKeys[normalized] {
		return true
	}
	for _, part := range strings.Split(normalized, "_") {
		if sensitiveKeys[part] {
			return true
		}
	}
	return false
}

func nullString(value *string) sql.NullString {
	if value == nil {
		return sql.NullString{}
	}
	return sql.NullString{String: *value, Valid: true}
}
